### ISSUE TITLE
[question] Does removing MAX_PATH Windows limitation work as expected?

### ISSUE THREAD

**Issue by daywhirls** [conan-io/conan#12738]
Hello,

I am seeing an issue on Windows when trying to package long-named folders after enabling long paths at the system level. I have some prebuilt binaries that I am trying to package, and during a `conan export-pkg` or `conan create` invoke, I am left with some 300+ length paths when it tries to copy files to the package folder. When `package()` runs, I get the following error:
```
ERROR: PACKAGE_NAME/PACKAGE_VERSION@USER/CHANNEL: Error in package() method, line 44
        self.copy("*")
        FileNotFoundError: [Errno 2] No such file or directory: 'C:\\Users\\me\\.conan\\data\\PACKAGE_NAME\\PACKAGE_VERSION\\USER\\CHANNEL\\package\\<LONG_HASH>\\very_long_path\file.ext'
```
It definitely seems to be a MAX_PATH issue because it copies files under the MAX_PATH limit just fine, and if I adjust the path to make it slightly smaller then it will continue a little further in the copy process before it finds the next too-long path. I am wondering why I'm having this issue after I've enabled long paths in the registry.

There is a note in the documentation [here](https://docs.conan.io/en/latest/reference/conanfile/attributes.html#short-paths) that states:
```
With the path limit removed both on the OS and on Python, the short_paths functionality becomes unnecessary, and can be disabled explicitly through the CONAN_USER_HOME_SHORT environment variable
```

I am confused with the statement "both on the OS and on Python," because my searches indicate that both OS and Python Installer set and use the same registry value for this. I have followed the instructions in the Microsoft documentation to remove the MAX_PATH limit. I have modified the `LongPathsEnabled` registry key, and I also enabled the group policy for good measure via Computer Configuration > Administrative Templates > System > Filesystem > Enable Win32 long paths. After rebooting the machine, it seems that Conan/Python is still unable to copy files that are past the 260 limit. I have also tried setting `CONAN_USER_HOME_SHORT` environment variable to `None` but it seems to have no effect. 

Am I missing something, or could this be a real issue? I am using **Conan 1.54** and **Python 3.10.5**.  I verified with Python outside of Conan that using `shutil.copyfile()` with extra long paths succeeds just fine, and also verified Python knows long paths are enabled with `RtlAreLongPathsEnabled()` function. So while it sounds like a Python/OS problem initially, I am led to suspect something Conan-specific could be at play here.

Thanks in advance for your time and insight. I would very much like to avoid using `short_paths` if possible, as it is pretty chaotic and annoying to navigate the cache, so hoping I am missing something obvious!

### Have you read the CONTRIBUTING guide?

- [X] I've read the CONTRIBUTING guide

**Comment by memsharded** [comment]
Hi @daywhirls 

Thanks very much for your report.
I am having a look, and I don't find anything in the ``self.copy()`` that would be limiting this.
But the ``self.copy()`` does a bit too much, and it has been superseded by a ``copy(self...)`` (``from conan.tools.files import copy``) that is simpler. (You might find this recommendation in this guide: https://docs.conan.io/en/latest/conan_v2.html)

Could you please try with that one to see if the problem persists?

**Comment by daywhirls** [comment]
HI @memsharded,

Thank you for the quick response! I did verify that the new `copy(self...)` function produces the same error. 

I have created a minimum reproducible example with the following batch script. I have tested this on two different machines, though they are both identical, running Windows 10 22H2 (19045.2251). 

```
mkdir C:\maxpathtest && cd C:\maxpathtest
conan new maxpathtest/1.0 --bare
mkdir aaaaaaaaaa\bbbbbbbbbb\cccccccccc\dddddddddd\eeeeeeeeee\ffffffffff\gggggggggg\hhhhhhhhhh\iiiiiiiiii\jjjjjjjjjj\kkkkkkkkkk\llllllllll\mmmmmmmmmm\nnnnnnnnnn\oooooooooo\pppppppppp
echo hello > aaaaaaaaaa\bbbbbbbbbb\cccccccccc\dddddddddd\eeeeeeeeee\ffffffffff\gggggggggg\hhhhhhhhhh\iiiiiiiiii\jjjjjjjjjj\kkkkkkkkkk\llllllllll\mmmmmmmmmm\nnnnnnnnnn\oooooooooo\pppppppppp\test.h
conan export-pkg . user/channel
```
The output of the export-pkg command is: 
```
Exporting package recipe
maxpathtest/1.0@user/channel: A new conanfile.py version was exported
maxpathtest/1.0@user/channel: Folder: C:\Users\<me>\.conan\data\maxpathtest\1.0\user\channel\export
maxpathtest/1.0@user/channel: Using the exported files summary hash as the recipe revision: 5f237d1eb01a58a1a4b8bd3d3f6dc802
maxpathtest/1.0@user/channel: Exported revision: 5f237d1eb01a58a1a4b8bd3d3f6dc802
maxpathtest/1.0@user/channel: Forced build from source
Packaging to 3fb49604f9c2f729b85ba3115852006824e72cab
maxpathtest/1.0@user/channel: Generating the package
maxpathtest/1.0@user/channel: Package folder C:\Users\<me>\.conan\data\maxpathtest\1.0\user\channel\package\3fb49604f9c2f729b85ba3115852006824e72cab
maxpathtest/1.0@user/channel: Calling package()
ERROR: maxpathtest/1.0@user/channel: Error in package() method, line 15
        self.copy("*")
        FileNotFoundError: [Errno 2] No such file or directory: 'C:\\Users\\<me>\\.conan\\data\\maxpathtest\\1.0\\user\\channel\\package\\3fb49604f9c2f729b85ba3115852006824e72cab\\aaaaaaaaaa\\bbbbbbbbbb\\cccccccccc\\dddddddddd\\eeeeeeeeee\\ffffffffff\\gggggggggg\\hhhhhhhhhh\\iiiiiiiiii\\jjjjjjjjjj\\kkkkkkkkkk\\llllllllll\\mmmmmmmmmm\\nnnnnnnnnn\\oooooooooo\\pppppppppp\\test.h'
```

**Comment by memsharded** [comment]
Thanks for the repro case, it really helps.

I am able to reproduce, and if I set the CONAN_VERBOSE_TRACEBACK=1, I am getting:

```
maxpathtest/1.0@user/channel: Calling package()
ERROR: Traceback (most recent call last):
  File "c:\users\memsharded\conanws\conan_src\conans\errors.py", line 34, in conanfile_exception_formatter
    yield
  File "c:\users\memsharded\conanws\conan_src\conans\client\conanfile\package.py", line 52, in _call_package
    conanfile.package()
  File "C:\Users\memsharded\.conan\data\maxpathtest\1.0\user\channel\export\conanfile.py", line 16, in package
    copy(self, "*", src=self.source_folder, dst=self.package_folder)
  File "c:\users\memsharded\conanws\conan_src\conan\tools\files\copy_pattern.py", line 42, in copy
    copied_files = _copy_files(files_to_copy, src, dst, keep_path)
  File "c:\users\memsharded\conanws\conan_src\conan\tools\files\copy_pattern.py", line 127, in _copy_files
    shutil.copy2(abs_src_name, abs_dst_name)
  File "c:\ws\python\python36\lib\shutil.py", line 263, in copy2
    copyfile(src, dst, follow_symlinks=follow_symlinks)
  File "c:\ws\python\python36\lib\shutil.py", line 121, in copyfile
    with open(dst, 'wb') as fdst:
FileNotFoundError: [Errno 2] No such file or directory: 'C:\\Users\\memsharded\\.conan\\data\\maxpathtest\\1.0\\user\\channel\\package\\3fb49604f9c2f729b85ba3115852006824e72cab\\aaaaaaaaaa\\bbbbbbbbbb\\cccccccccc\\dddddddddd\\eeeeeeeeee\\ffffffffff\\gggggggggg\\hhhhhhhhhh\\iiiiiiiiii\\jjjjjjjjjj\\kkkkkkkkkk\\llllllllll\\mmmmmmmmmm\\nnnnnnnnnn\\oooooooooo\\pppppppppp\\test.h'
```

This is expected, as I don't have the long paths enabled, the path that you created contains around 190 chars, totally ok, but the path in the Conan cache has > 279, and then it makes sense that it fails.

If your system is allowing long paths, the first path in your repro case that you are created should already be >279 chars. If I try to create that folder and file in the first place, the OS will not allow me, giving me an error message.

Can you please try to reproduce with the first path using a >280 chars one?

**Comment by daywhirls** [comment]
Okay, I tried reproducing while increasing the first path past MAX_PATH length, which created just fine on my computer:
```
mkdir C:\maxpathtest && cd C:\maxpathtest
conan new maxpathtest/1.0 --bare
mkdir aaaaaaaaaa\bbbbbbbbbb\cccccccccc\dddddddddd\eeeeeeeeee\ffffffffff\gggggggggg\hhhhhhhhhh\iiiiiiiiii\jjjjjjjjjj\kkkkkkkkkk\llllllllll\mmmmmmmmmm\nnnnnnnnnn\oooooooooo\pppppppppp\qqqqqqqqqq\rrrrrrrrrr\ssssssssss\tttttttttt\uuuuuuuuuu\vvvvvvvvvv\wwwwwwwwww\xxxxxxxxxx\yyyyyyyyyy\zzzzzzzzzz
echo hello > aaaaaaaaaa\bbbbbbbbbb\cccccccccc\dddddddddd\eeeeeeeeee\ffffffffff\gggggggggg\hhhhhhhhhh\iiiiiiiiii\jjjjjjjjjj\kkkkkkkkkk\llllllllll\mmmmmmmmmm\nnnnnnnnnn\oooooooooo\pppppppppp\qqqqqqqqqq\rrrrrrrrrr\ssssssssss\tttttttttt\uuuuuuuuuu\vvvvvvvvvv\wwwwwwwwww\xxxxxxxxxx\yyyyyyyyyy\zzzzzzzzzz\test.h
conan export-pkg . user/channel
```

The output for `export-pkg` is:
```
C:\maxpathtest>conan export-pkg . user/channel
Exporting package recipe
maxpathtest/1.0@user/channel: A new conanfile.py version was exported
maxpathtest/1.0@user/channel: Folder: C:\Users\<me>\.conan\data\maxpathtest\1.0\user\channel\export
maxpathtest/1.0@user/channel: Using the exported files summary hash as the recipe revision: 60d29eb2230bf03cf704817bfc1026f4
maxpathtest/1.0@user/channel: Removing the local binary packages from different recipe revisions
maxpathtest/1.0@user/channel: Exported revision: 60d29eb2230bf03cf704817bfc1026f4
maxpathtest/1.0@user/channel: Forced build from source
Packaging to 3fb49604f9c2f729b85ba3115852006824e72cab
maxpathtest/1.0@user/channel: Generating the package
maxpathtest/1.0@user/channel: Package folder C:\Users\<me>\.conan\data\maxpathtest\1.0\user\channel\package\3fb49604f9c2f729b85ba3115852006824e72cab
maxpathtest/1.0@user/channel: Calling package()
maxpathtest/1.0@user/channel package(): Packaged 1 '.py' file: conanfile.py
maxpathtest/1.0@user/channel: Package '3fb49604f9c2f729b85ba3115852006824e72cab' created
maxpathtest/1.0@user/channel: Created package revision 8f0c81908cd81e332e81c6f5fb60f7e6
```
I'm not sure why, but the command succeeds yet doesn't copy anything to the package directory. In the previous example, the `package` folder would have the `C:\Users\<me>\.conan\data\maxpathtest\1.0\user\channel\package\3fb49604f9c2f729b85ba3115852006824e72cab\aaaaaaaaaa\bbbbbbbbbb\cccccccccc\dddddddddd\eeeeeeeeee\ffffffffff\gggggggggg\hhhhhhhhhh\iiiiiiiiii\jjjjjjjjjj\kkkkkkkkkk` before it halts trying to copy the next folder due to hitting max path. This time it copied nothing, so the package folder only contains the conanfile, conaninfo, and conanmanifest files. I am not sure how to properly reproduce because of this.

> the path that you created contains around 190 chars, totally ok, but the path in the Conan cache has > 279, and then it makes sense that it fails.
> If your system is allowing long paths, the first path in your repro case that you are created should already be >279 chars

The issue I am experiencing is how the first example works, where the files I want to package are all under MAX_PATH length normally, but they become > MAX_PATH during the packaging process and throw the error above due to how long the Conan cache path is. I expected that removing the MAX_PATH limit would handle situations where copying the "before" path is fine but the "after" path is too long, but when I try this, the result looks the same as if the MAX_PATH limit was not removed.

**Comment by daywhirls** [comment]
It seems that I can reproduce the same error in Python alone using the copy function used by Conan:
```
import shutil
shutil.copy2("C:\\maxpathtest\\aaaaaaaaaa\\bbbbbbbbbb\\cccccccccc\\dddddddddd\\eeeeeeeeee\\ffffffffff\\gggggggggg\\hhhhhhhhhh\\iiiiiiiiii\\jjjjjjjjjj\\kkkkkkkkkk\\llllllllll\\mmmmmmmmmm\\nnnnnnnnnn\\oooooooooo\\pppppppppp\\qqqqqqqqqq\\rrrrrrrrrr\\ssssssssss\\tttttttttt\\uuuuuuuuuu\\vvvvvvvvvv\\wwwwwwwwww\\xxxxxxxxxx\\yyyyyyyyyy\\zzzzzzzzzz\\test.h", "C:\\Users\\daywhirls\\.conan\\data\\maxpathtest\\1.0\\user\\channel\\package\\3fb49604f9c2f729b85ba3115852006824e72cab\\aaaaaaaaaa\\bbbbbbbbbb\\cccccccccc\\dddddddddd\\eeeeeeeeee\\ffffffffff\\gggggggggg\\hhhhhhhhhh\\iiiiiiiiii\\jjjjjjjjjj\\kkkkkkkkkk\\llllllllll\\mmmmmmmmmm\\nnnnnnnnnn\\oooooooooo\\pppppppppp\\qqqqqqqqqq\\rrrrrrrrrr\\ssssssssss\\tttttttttt\\uuuuuuuuuu\\vvvvvvvvvv\\wwwwwwwwww\\xxxxxxxxxx\\yyyyyyyyyy\\zzzzzzzzzz\\test.h")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "bin\Python\3.10.5-win\lib\shutil.py", line 434, in copy2
    copyfile(src, dst, follow_symlinks=follow_symlinks)
  File "bin\Python\3.10.5-win\lib\shutil.py", line 256, in copyfile
    with open(dst, 'wb') as fdst:
FileNotFoundError: [Errno 2] No such file or directory: 'C:\\Users\\daywhirls\\.conan\\data\\maxpathtest\\1.0\\user\\channel\\package\\3fb49604f9c2f729b85ba3115852006824e72cab\\aaaaaaaaaa\\bbbbbbbbbb\\cccccccccc\\dddddddddd\\eeeeeeeeee\\ffffffffff\\gggggggggg\\hhhhhhhhhh\\iiiiiiiiii\\jjjjjjjjjj\\kkkkkkkkkk\\llllllllll\\mmmmmmmmmm\\nnnnnnnnnn\\oooooooooo\\pppppppppp\\qqqqqqqqqq\\rrrrrrrrrr\\ssssssssss\\tttttttttt\\uuuuuuuuuu\\vvvvvvvvvv\\wwwwwwwwww\\xxxxxxxxxx\\yyyyyyyyyy\\zzzzzzzzzz\\test.h'
```
But that error seems to be happening because the `aaaaaaaaaa\\bbbbbbbbbb\\cccccccccc\\dddddddddd\\eeeeeeeeee\\ffffffffff\\gggggggggg\\hhhhhhhhhh\\iiiiiiiiii\\jjjjjjjjjj\\kkkkkkkkkk\\llllllllll\\mmmmmmmmmm\\nnnnnnnnnn\\oooooooooo\\pppppppppp\\qqqqqqqqqq\\rrrrrrrrrr\\ssssssssss\\tttttttttt\\uuuuuuuuuu\\vvvvvvvvvv\\wwwwwwwwww\\xxxxxxxxxx\\yyyyyyyyyy\\zzzzzzzzzz` folders don't exist in the destination path. So if I do this:
```
os.makedirs('C:/Users/daywhirls/.conan/data/maxpathtest/1.0/user/channel/package/3fb49604f9c2f729b85ba3115852006824e72cab/aaaaaaaaaa/bbbbbbbbbb/cccccccccc/dddddddddd/eeeeeeeeee/ffffffffff/gggggggggg/hhhhhhhhhh/iiiiiiiiii/jjjjjjjjjj/kkkkkkkkkk/llllllllll/mmmmmmmmmm/nnnnnnnnnn/oooooooooo/pppppppppp/qqqqqqqqqq/rrrrrrrrrr/ssssssssss/tttttttttt/uuuuuuuuuu/vvvvvvvvvv/wwwwwwwwww/xxxxxxxxxx/yyyyyyyyyy/zzzzzzzzzz'
shutil.copy2("C:\\maxpathtest\\aaaaaaaaaa\\bbbbbbbbbb\\cccccccccc\\dddddddddd\\eeeeeeeeee\\ffffffffff\\gggggggggg\\hhhhhhhhhh\\iiiiiiiiii\\jjjjjjjjjj\\kkkkkkkkkk\\llllllllll\\mmmmmmmmmm\\nnnnnnnnnn\\oooooooooo\\pppppppppp\\qqqqqqqqqq\\rrrrrrrrrr\\ssssssssss\\tttttttttt\\uuuuuuuuuu\\vvvvvvvvvv\\wwwwwwwwww\\xxxxxxxxxx\\yyyyyyyyyy\\zzzzzzzzzz\\test.h", "C:\\Users\\daywhirls\\.conan\\data\\maxpathtest\\1.0\\user\\channel\\package\\3fb49604f9c2f729b85ba3115852006824e72cab\\aaaaaaaaaa\\bbbbbbbbbb\\cccccccccc\\dddddddddd\\eeeeeeeeee\\ffffffffff\\gggggggggg\\hhhhhhhhhh\\iiiiiiiiii\\jjjjjjjjjj\\kkkkkkkkkk\\llllllllll\\mmmmmmmmmm\\nnnnnnnnnn\\oooooooooo\\pppppppppp\\qqqqqqqqqq\\rrrrrrrrrr\\ssssssssss\\tttttttttt\\uuuuuuuuuu\\vvvvvvvvvv\\wwwwwwwwww\\xxxxxxxxxx\\yyyyyyyyyy\\zzzzzzzzzz\\test.h")
```
Then everything works. Could it be that Conan is having trouble making the directories past MAX_PATH before the copy is performed?

**Comment by memsharded** [comment]
Good investigation.
I am having a look, and I can't find a reason why Conan would fail to create the directory, actually it uses also internally the same ``os.makedirs()`` Python function.

Maybe you can omit the ``copy()`` (or the ``self.copy()``) helpers and use ``shutil`` and ``os.makedirs`` directly in the ``package()`` method. There is nothing really special in the ``copy()`` function, and you have the necessary folders to reference : ``self.source_folder``, ``self.package_folder``, and then see if it works? I assume that you are installing Conan with the ``pip install``? Other installers might not be using your system Python, but a bundled Python.

**Comment by daywhirls** [comment]
Hi @memsharded, 

I spent some more time on this to try and understand what is happening, and I think that either I am fundamentally misunderstanding something crucial about how bypassing MAX_PATH on Windows works, or Python actually is not able to handle this functionality.

As an experiment I tried manually creating a directory past MAX_PATH like above prior to the file being copied, and everything works. But it only works when I manually hardcode every single path (impractical for large folder trees). I tried to write Python code that traverses a folder structure to populate a list that can be used to automatically make any directories that may be past MAX_PATH, but I could not find a single solution that didn't run into the same exact errors mentioned above. Every Python function I tried across several modules (glob, shutil, etc.) had the same limitation of reporting `No such file or directory` as soon as it tried to look for something past the MAX_PATH amount of characters.

So either I have misunderstood something along the way, or Python cannot properly handle the situation where a file is being copied from a `< MAX_PATH` length to a folder that is `> MAX_PATH` length, seemingly due to being unable to create the missing folder structure past the MAX_PATH mark. If this limitation is true, then Conan will fail in situations where a deep folder structure (< MAX_PATH) being copied to a package inside the Conan cache exceeds MAX_PATH length due to the Conan cache package directory length, even if MAX_PATH bypass is enabled on the Windows system.

If anyone else can confirm the same findings or explain that I'm wrong, that would be greatly appreciated!

As this doesn't technically appear to be a Conan-specific problem, I'm okay with this issue being closed, but it does seem worth getting additional confirmation since this could be misleading in the Conan documentation if my problem is legitimate.

**Comment by memsharded** [comment]
Thanks very much @daywhirls for the deep investigation.
The truth is that we implemented the ``short_paths`` feature in Conan 1.X because relying on long paths in Windows was always problematic, it never worked smoothly. It is very possible that there are still some limitations inside Python, as we are aware those limitations still happen in different tools like build systems and compilers.

Furthermore, this is also one of the reasons we have done a great effort in the new Conan 2.0 to have the paths shortened with a new cache structure, so paths will be always shorter, and do not rely on the ``short_paths`` feature either. The intention is that Conan cache paths are no longer an issue, and then there should be nothing to explain in the docs. So I would suggest to close this issue as addressed by 2.0.

**Comment by daywhirls** [comment]
That is great news to hear about the Conan 2.0 improvements shortening cache paths without the need for `short_paths`! I am very happy to hear that and look forward to using it.

I will close this issue now - thanks again for all of your insight!
